lib/typing/mongo/bson.go

```diff
@@ -32,133 +27,101 @@ func JSONEToMap(val []byte) (map[string]any, error) {
+		value, err := bsonValueToGoValue(elem.Value)
+		if err != nil {
+			return nil, err
+		}
 
-	ints, err := strconv.Atoi(fmt.Sprint(val))
-	if err != nil {
-		return err
+		result[elem.Key] = value
 	}
-
-	t := time.Unix(0, int64(ints)*1000000).UTC()
-	return vw.WriteString(t.Format(ext.ISO8601))
+	return result, nil
 }
 
-func objectIDEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
-	if !val.IsValid() || val.Type() != tOID {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue", Types: []reflect.Type{tOID}, Received: val}
-	}
+func bsonArrayToSlice(arr bson.A) ([]any, error) {
+	result := make([]any, len(arr))
+	for i, elem := range arr {
+		value, err := bsonValueToGoValue(elem)
+		if err != nil {
+			return nil, err
+		}
 
-	s, isOk := val.Interface().(primitive.ObjectID)
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not objectID", Types: []reflect.Type{tOID}, Received: val}
+		result[i] = value
 	}
-
-	return vw.WriteString(s.Hex())
+	return result, nil
 }
 
-func binaryEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
-	if !val.IsValid() || val.Type() != tBinary {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	s, isOk := val.Interface().(primitive.Binary)
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not Binary", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	switch s.Subtype {
+func bsonBinaryValueToMap(value primitive.Binary) (any, error) {
+	switch value.Subtype {
 	case
 		bson.TypeBinaryUUIDOld,
 		bson.TypeBinaryUUID:
-		parsedUUID, err := uuid.FromBytes(s.Data)
+		parsedUUID, err := uuid.FromBytes(value.Data)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		return vw.WriteString(parsedUUID.String())
+
+		return parsedUUID.String(), nil
 	default:
-		return vw.WriteBinaryWithSubtype(s.Data, s.Subtype)
+		return map[string]any{
+			"$binary": map[string]any{
+				"base64":  base64.StdEncoding.EncodeToString(value.Data),
+				"subType": fmt.Sprintf("%02x", value.Subtype),
+			},
+		}, nil
 	}
 }
 
-func timestampEncodeValue(_ bsoncodec.EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
-	if !val.IsValid() || val.Type() != tTimestamp {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue", Types: []reflect.Type{tBinary}, Received: val}
-	}
-
-	s, isOk := val.Interface().(primitive.Timestamp)
-	if !isOk {
-		return bsoncodec.ValueEncoderError{Name: "ObjectIDEncodeValue not Binary", Types: []reflect.Type{tBinary}, Received: val}
-	}
+func bsonValueToGoValue(value any) (any, error) {
+	switch v := value.(type) {
+	case primitive.DateTime:
+		return time.Unix(0, int64(v)*int64(time.Millisecond)).UTC().Format(ext.ISO8601), nil
+	case primitive.ObjectID:
+		return v.Hex(), nil
+	case primitive.Binary:
+		return bsonBinaryValueToMap(v)
+	case primitive.Decimal128:
```

Comment:
Do we need to handle decimals that between 64 and 128 bits in size?

Reply:
wdym?

Reply:
If it exceeds float64, we'll return it as a string

Reply:
Won't we get a `value out of range` for numbers too big to be represented by float64, like `1.7E+309`?

Reply:
I expected to - but we didn't.

See: https://github.com/artie-labs/transfer/blob/15cabd47fa783ce5e7a0b967d399857e5b6aa26e/lib/typing/mongo/bson_test.go#L216-L221

Reply:
I think that's just because the `fmt.Sprint()` is truncating the value to `1.7976931348623157e+308` when it converts to exponential notation and that value is representable by a float64. If you change it to `primitive.ParseDecimal128("1.7e+309")` you do get an out of range error.